New `KeyEvent` handlers and main event loop rewrite

gedankenexperimenter opened the pull request:
This PR includes changes from several others (particularly #904, but also #905 and #913). I'm sorry for the clutter this is causing, but disentangling those other changes to minimize the amount of code that changed here just wasn't worth the effort.

This PR _shouldn't_ break user sketches that use only core Kaleidoscope plugins. It's possible that it does in its current form—not every core plugin has testcases yet—but I'm pretty sure we can keep those sketches working. Users of custom Macros are somewhat likely to need to make adjustments, and third-party plugins may break (though I have done a fair amount of work to keep even those functional, this changes the plugin API in substantial ways). The good news is that the new version of most plugins is simpler and easier to write (and get correct) than the old version was. See SpaceCadet in this PR as an example of that.

With all those caveats out of the way, I will say that this PR fixes a number of bugs in the current version of Kaleidoscope, and passes all existing testcases for me, including a number of new ones. I would not merge this without adding a bunch of testcases for plugins that don't have any yet. That's probably the biggest hole.

The most important places to look if you're investigating the new code are `src/kaleidoscope/Runtime.{h,cpp}` and `docs/api-reference/event-handler-hooks.md` (still not quite complete, but hopefully helpful).

Plugins that may be particularly useful to look at include Qukeys, SpaceCadet, TapDance, and OneShot.

I plan to add two plugins soon that are very difficult to write for current master, but almost trivial after these changes: Unshifter (like TopsyTurvy + ShapeShifter, but even more flexible) and AutoShift.

More documentation is needed, of course. I haven't yet added anything to UPGRADING.md or NEWS.md, but I will.

I would be delighted to answer anyone's questions about this PR, and if anyone wants to try it out in order to get access to any of the bug fixes or new features included, I should be available to help with almost anything except how to build firmware on Windows.

## Testing

Here's a checklist of plugins, indicating which ones have been tested (currently incomplete):

- [ ] Colormap
- [x] Cycle
- [ ] DynamicMacros
- [ ] DynamicTapDance
- [ ] EEPROM-*
- [x] Escape-OneShot
- [ ] FingerPainter
- [x] GhostInTheFirmware
- [x] Heatmap
- [x] IdleLEDs
- [x] Leader
- [x] LED-ActiveModColor
- [x] LED-AlphaSquare
- [x] LED-Stalker
- [x] LED-Wavepool
- [x] LEDEffect-*
- [x] Macros
- [x] MagicCombo
- [x] MouseKeys
- [x] OneShot
- [x] Qukeys
- [x] Redial
- [x] ShapeShifter
- [x] SpaceCadet
- [ ] Steno (GeminiPR)
- [x] Syster
- [x] TapDance
- [x] TopsyTurvy
- [x] Turbo
- [ ] TypingBreaks
- [x] WinKeyToggle

Hardware:

- [ ] Dygma Raise
- [x] Ergodox EZ
- [ ] gHeavy Butterstick
- [ ] gHeavy Faunchpad
- [x] KBDFans KBD4x
- [x] Keyboardio Atreus
- [x] Keyboardio Model01
- [ ] OLKB Planck
- [x] Technomancy Atreus
- [x] SOFTHRUF Splitography

Other things that also need testing:

- [ ] Chrysalis (this should probably get broken up into different items).

## Documentation

Some docs are written, at least in rough draft form. Here's a checklist:

- [x] UPGRADING.md
- [ ] NEWS.md
- [x] plugin-handler-hooks.md
- [ ] More detailed notes on plugin interactions (init order, mostly)
- [ ] Deprecation dates (if and when the decision is made to merge)

## Precursor PRs

Closes #904 
Closes #905 
Closes #913 

## Issues addressed

Fixes #407
Fixes #484
Fixes #571
Fixes #623
Fixes #673
Fixes #806 
Fixes #882
Fixes #894
Fixes #895
Fixes #896
Fixes #922 
Fixes #985 
Fixes #980 
Fixes #990 

Closes #568
Closes #906 
Closes #907
Closes #419
Closes #414 (debatably)
Closes #399

obra commented:
Thanks for submitting this. I know you've been working toward it for quite a while. And it ended up with a very auspicious PR #

gedankenexperimenter commented:
> Thanks for submitting this. I know you've been working toward it for quite a while. And it ended up with a very auspicious PR #

Heh. I honestly didn't even notice the PR # while I was submitting it.

gedankenexperimenter commented:
I'm continuing to work on this branch. LED modes are probably broken to some extent at the moment, but I'm working on it. I'm thinking of adding another event handler: `onSyncLeds()`, to let plugins override the current LED mode more efficiently, just before the LEDs get updated, rather than every cycle.

gedankenexperimenter commented:
This PR is now substantially more complete, with a number of deprecations, a couple of significant plugin rewrites (Macros in particular), and all existing plugins converted to the new event handlers. There's still major work to be done reorganizing the commits, and—most important—the addition of testcases for lots of plugins.

gedankenexperimenter commented:
I've done a major rewrite of the commit history, and this PR is getting close to being presentable. It still needs more and better docs, and a bunch of testcases, as well as some testing on hardware (especially for LEDs). No doubt there are still some bugs to work out, of course.

I look forward to discussing these changes when you have the time.

gedankenexperimenter commented:
Update: We've been working through some bugs, and improving the helper class infrastructure, probably working toward some future potential improvements (a global event queue is the big one I'm thinking about). While doing that, plugin initialization order and related issues came up again, and I want to record some insights here lest I forget them.

First, there was an issue where two plugins that both implement `onPhysicalKeyEvent()` (Qukeys & AutoShift) were interacting in a detrimental way. We had set up AutoShift first in `KALEIDOSCOPE_INIT_PLUGINS()`, but found that it wasn't working. The real reason turned out to be that Qukeys wasn't using a `KeyEventTracker` to properly ignore events that it had already seen (fixed now), but it made me realize something crucial:

Qukeys ignores the `Key` value of events it receives, because for anything except the first event in the queue, it's not reliable, because that first event might resolve to a layer change (most often a layer shift, if it's held long enough). The problem is that if another plugin processes the event before Qukeys, _and changes the event's `Key` value_, that change will get discarded. And if Qukeys resolves a prior event to a layer shift (or, more generally any alternate value), the only way to resolve that altered `Key` value correctly would be to "rewind" that plugin to a time before it first encountered the event, and have it start over. Needless to say, that's far too impractical to even consider.

I have two ideas for ways to address this concern. First, I want to acknowledge that it would be cute to have a Qukeys key resolve to a SpaceCadet key which resolves to a TapDance key, which then gets AutoShifted, but there's vanishingly small utility in having more than one of these plugins involved in the same keymap entry. We could just simplify by changing the contract for `onPhysicalKeyEvent()` handlers from "don't process the same event twice" to "don't call `handlePhysicalKeyEvent()`—alway skip straight to `handleKeyEvent()`". If all the `onPhysicalKeyEvent()` plugins abide by that rule, a few possibilities would be closed (e.g. Qukeys modifier on hold, TapDance on tap), but things would be simpler. We could simply revise the existing rule to explicitly give plugins the option to skip straight to `handleKeyEvent()`.

The other possibility that might help is an event queue that stores the (pending) `Key` values of an event, in addition to its other details. Sometimes (particularly in the case of intervening layer shifts as described above) this could backfire or require substantial extra code, but it might help more than it hurts.

gedankenexperimenter commented:
Right now we have a new AutoShift plugin in this branch. @obra, would you prefer that to be broken out into a separate PR?

algernon commented:
Before I start diving deeper, I'll share my opinion on some of the latest, remaining questions:

> I have two ideas for ways to address this concern. First, I want to acknowledge that it would be cute to have a Qukeys key resolve to a SpaceCadet key which resolves to a TapDance key, which then gets AutoShifted, but there's vanishingly small utility in having more than one of these plugins involved in the same keymap entry. We could just simplify by changing the contract for `onPhysicalKeyEvent()` handlers from "don't process the same event twice" to "don't call `handlePhysicalKeyEvent()`—alway skip straight to `handleKeyEvent()`". If all the `onPhysicalKeyEvent()` plugins abide by that rule, a few possibilities would be closed (e.g. Qukeys modifier on hold, TapDance on tap), but things would be simpler. We could simply revise the existing rule to explicitly give plugins the option to skip straight to `handleKeyEvent()`.

I would strongly support this option. First of all, because it is simpler, smaller, faster, easier. Second, because while - as you said - it would be cute to have some complex interactions between certain plugins, my gut feeling is that doing so would be a very, very rare case. Not something we should be worrying about right now. I mean, those cases you listed, they don't work (either at all, or if they do, not reliably) now either, so if we go with this first option, we do not introduce a regression.

If - and when - a sizable demand manifests for such interactions, we can always reconsider later, and see if we can find a way to make them work. Either with your second option, or some other way. But for now, I'd stick to the simpler way, even if that's considerably less flexible.

gedankenexperimenter commented:
> Second, because while - as you said - it would be cute to have some complex interactions between certain plugins, my gut feeling is that doing so would be a very, very rare case.

As a devil's advocate argument, I think there's one case that could be very useful, and relies on a combination of two of the aforementioned plugins: Qukeys -> TapDance (or possibly SpaceCadet -> TapDance), which would allow one behaviour for a key on a hold, and a set of alternates based on the number of taps. I believe this can be done reliably, and offers genuine utility.

gedankenexperimenter commented:
MouseKeys are now fully functional again (tested on my Model01).

algernon commented:
## GhostInTheFirmware

This does not seem to work with this PR. But it doesn't seem to work on master, either. Or I forgot what it is supposed to do. Will test it on the Model01 later with its example sketch.

## Turbo

Likewise, this does not appear to work with the PR, but it does work on master. I'll investigate further once I'm done testing the rest.

gedankenexperimenter commented:
Both of those were non-trivial conversions, so I'm not too surprised. I'm confident that I can get Turbo working today, at least (unless I'm too busy with tax returns). GhostInTheFirmware is definitely less clear to me.

algernon commented:
A'ight, so, I'd like to share my general opinion on the PR and the direction.

I was skeptical in the beginning, because it is a _huge_ change, both in the sense of code changed, and in the sense of architecture and behaviour too. Changing the key event  handlers from running each cycle to be event-driven was strange at first, and having worked with the current behaviour for so long, I wasn't sold on the idea at first. However, the more I thought about it, the more I realized that an event-driven architecture actually makes _more_ sense. USB sends reports when there's a change, after all, so it makes sense to handle key events when there _are_ key events. I came to really like what I'm seeing here.

Another great thing is the separation of physical and virtual key events - that makes a whole lot of things much, much simpler. I love it. I do believe we should keep the `onKeyswitchEvent` name, even if that'd be a bit confusing, but more on that - and naming, in general - later. Now that I see the whole picture, this is an entirely different beast, and one I _like_.

Furthermore, I can't stop smiling at `KeyEventTracker`, which is a way smarter solution for the problem of avoiding processing the same event twice than the `| INJECTED` and similar hacks we've been using so far. A little more expensive one, but unlike the tricks we used to date, rather more reliable, and better documented too.

The comments in the code are great, by the way, they really help understanding the code.

All in all, unlike in previous iterations, I'm very much in favour of getting this in. We will need to test the remaining plugins, and fix any regressions (unless we decide that a regression is something we can live with, primarily with regards to how certain plugins can interact), and possibly have a little chat to figure out better names for some of the things. In particular, `onPhysicalKeyEvent` and `onKeyEvent` could use better names, I think. Not 100% sold on `KeyEventTracker` or `KeyAddrEventQueue` either. I could also imagine a better name for `LiveKeys`/`Kaleidoscope::live_keys[]` too.

I will be leaving in-line comments here and there in the next few hours, most likely. But having combed through the PR a few times now, and seeing how some of the plugins got simpler, and more understandable too, I'm enthusiastic.

gedankenexperimenter commented:
Here's a big question, that I think is important to discuss before this gets finalized: Should `KeyEvent` be a fully-encapsulated class instead of a POD struct?

You probably noticed that I've been unable to resist the urge, when doing substantial rewrites of certain plugins, to deprecate public variables, especially ones that don't conform to the naming conventions in the style guide. I've waffled on making `KeyEvent` keeping its data accessible only via methods. I do like the simplicity of just writing `event.key` and `event.state`, but it's obviously not as robust from a maintenance perspective as `event.key()` and `event.state()`, especially when setting values (`event.key = Key_X` is more intuitive to comprehend than `event.setKey(Key_X)` (or `event.key() = Key_X` if there's a version of `KeyEvent::key()` that returns a mutable reference)). It doesn't seem likely that we would want the structure to change in any substantial way, and adding a new data member is not a problem, but there's always the possibility, however remote, that some name changes will be desired in the future. Also, if the data members are private, that makes it possible to control access in a more fine-grained way (from a function that takes a mutable `KeyEvent` reference, we could make it nevertheless impossible to alter the `id` value, for instance).

I'm still leaning towards it being a POD struct, to keep the code simpler, but there are some good arguments for making it a full-fledged class instead, and I want to make sure we discuss that option before merging this code, because it's much easier to do it now.

algernon commented:
I think a struct is fine for now. _If_ (and that's a big if) we ever feel the need to move towards something more complex that'd warrant accessors, we can still do so, most likely, with a little boilerplate and operator overloading.

gedankenexperimenter commented:
> I think a struct is fine for now. _If_ (and that's a big if) we ever feel the need to move towards something more complex that'd warrant accessors, we can still do so, most likely, with a little boilerplate and operator overloading.

Maybe it would be worthwhile to protect just the `id` member for now, since it generally shouldn't be altered by plugins (directly) anyway. For that matter, I think it might make sense to redo some of the `KeyEvent` constructors a bit, so that they never automatically increment the global event ID, but instead have the hardware class call a different function—let's call it `KeyEvent::next(key_addr, key_state)` to generate the new (physical) keyswitch events, just to further minimize the possibility of a plugin that creates lots of artificial events from increasing the event ID while bypassing `onPhysicalKeyEvent()` handlers, and causing those plugins to get confused about which events they've already processed.

algernon commented:
Yeah, protecting `id` makes sense.

gedankenexperimenter commented:
> Yeah, protecting `id` makes sense.

I'll make that change sometime this weekend.

gedankenexperimenter commented:
In making the change to protect `KeyEvent::id`, I noticed a minor flaw in the `EventTracker` system: for plugins using an event queue, if the `onPhysicalKeyEvent()` handler gets called with an `event.id` _that's in the queue_ (note: this should never happen), they should _not_ return `OK` because that event should still be waiting to be flushed from the queue. So I added some stuff to `KeyAddrEventQueue` to provide a `shouldAbort()` function so that if `EventTracker::shouldIgnore()` returns true, queueing plugins can return `ABORT` instead of `OK` for events that are in their queues. It does mean that those plugins need to update their event queues before calling `Runtime.handlePhysicalKeyEvent()` to avoid dropping the event, but that's not too bad.

gedankenexperimenter commented:
> We will need to test the remaining plugins, and fix any regressions (unless we decide that a regression is something we can live with, primarily with regards to how certain plugins can interact), and possibly have a little chat to figure out better names for some of the things. In particular, `onPhysicalKeyEvent` and `onKeyEvent` could use better names, I think. Not 100% sold on `KeyEventTracker` or `KeyAddrEventQueue` either. I could also imagine a better name for `LiveKeys`/`Kaleidoscope::live_keys[]` too.

Re: names: "Live keys" was something that @obra and I mentioned in a discussion about #904. I got the impression that it would be acceptable where "active keys" was not. I went with the operator overloads instead of `get*/set*` method names because those were also contentious, and I wanted it to be clear that it should be treated just like an array of `Key` objects, but addressed by `KeyAddr`, not by raw integers.

> I do believe we should keep the `onKeyswitchEvent` name, even if that'd be a bit confusing, but more on that - and naming, in general - later.

The reason I shied away from using the `onKeyswitchEvent()` name in favour of `onPhysicalKeyEvent()` is because the former would be the name that I'd expect people to look for out of inertia, regardless of whether or not there was a real key_switch_ involved. I like the name better than the more wordy one, but `onPhysicalKeyEvent()` has a lot less potential for misunderstanding. The function that should (almost) always be called by plugins and macros (and similar user code) is the one I've called `onKeyEvent()`, because it's more general (both physical and virtual events cause it to be called). I'd love to hear why you don't think that one is an appropriate name.

`KeyEventTracker` isn't something I have strong opinions about; I just wanted something reasonably descriptive, and it seemed to fit well enough.

`KeyAddrEventQueue` is a name that's been around for quite a while, but for most of its life, it existed only in Qukeys. Now there are a few other plugins that use it, too, so I moved it into the core, but didn't change the name. I named it that way because it doesn't include `Key` values (because they're currently unnecessary to the plugins that use it). I've been thinking quite a bit about the design of a global queue that would replace it, and that would include `Key` values as well, so would definitely get a different name. I'm fine with changing the name if we come up with something better.

gedankenexperimenter commented:
### Status update

- Most of the existing plugins have now been tested in one form or another. The biggest glaring hole is anything EEPROM-dependent, including Chrysalis compatibility, mostly due to my lack of familiarity with how any of that stuff works. Fortunately, the Focus code didn't really get touched at all by this change, so I hope it wasn't broken by it.

- Plugins haven't been testing in many combinations yet, either. That's another potential source of bugs, even though eliminating such bugs was one of the primary motivations for making these changes in the first place.

- Before merging all this into the master branch, I think it would be wise to get a few "power" users to test it out in real-world conditions to see if any problems emerge.

- This branch still includes the AutoShift plugin that @algernon and I put together. I think it should get a separate PR of its own, and I'm bothered by the fact that, because of commit squashing, there's a whole bunch of code that's attributed to me, but which was actually written by @algernon, and it makes me uncomfortable to be taking credit for someone else's work.

gedankenexperimenter commented:
I've removed AutoShift from this PR (will introduce that later), and removed placeholder deprecation dates & some trailing whitespace. I'm marking it ready for review.

gedankenexperimenter commented:
I've replaced `onPhysicalKeyEvent()` and `Runtime.handlePhysicalKeyEvent()` with `onKeyswitchEvent()` and `Runtime.handleKeyswitchEvent()`, as requested. I also added the aliases `Key_Masked` and `Key_Unbound` (though I want to go on record saying I think the latter is a misleading name). I think I've updated all code, comments, documentation, and commit messages to reflect these changes, and consider this PR to be ready to go (@algernon says that the Ergodox issue is not a blocker, and I don't have the hardware to test it myself).

I've been reasonably careful to use `Key_Masked` and `Key_NoKey` as correctly as possible, such that if the former gets redefined to a different value, things _should_ still work properly. There are some places where it's not 100% clear what to do if the two diverge, but it certainly seems to work for now.

As far as I know, there's no reason this PR shouldn't be merged as-is.

obra commented:
Do you have an alternate name you’d propose for Key_Unbound?

Thanks for all the work on this. 

> On May 25, 2021, at 10:29 PM, Michael Richters ***@***.***> wrote:
> 
> ﻿
> I've replaced onPhysicalKeyEvent() and Runtime.handlePhysicalKeyEvent() with onKeyswitchEvent() and Runtime.handleKeyswitchEvent(), as requested. I also added the aliases Key_Masked and Key_Unbound (though I want to go on record saying I think the latter is a misleading name). I think I've updated all code, comments, documentation, and commit messages to reflect these changes, and consider this PR to be ready to go ***@***.*** says that the Ergodox issue is not a blocker, and I don't have the hardware to test it myself).
> 
> I've been reasonably careful to use Key_Masked and Key_NoKey as correctly as possible, such that if the former gets redefined to a different value, things should still work properly. There are some places where it's not 100% clear what to do if the two diverge, but it certainly seems to work for now.
> 
> As far as I know, there's no reason this PR shouldn't be merged as-is.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

gedankenexperimenter commented:
I discovered a minor bug in the interaction with between the KeyEvent versions of TapDance and OneShot, where a pending TapDance key and a OneShot key are both resolved by the same event. In that case, the OneShot, which should apply only to the output of the TapDance key, applies to both that and the next keypress (the one that resolved the TapDance key). Since that's not very clear, here's an example:

1. Tap `OSM(LeftShift)`
2. Tap `TD(0)` (which outputs symbols `x`, `y`, or `z`, depending on the number of taps)
3. Tap `Key_A`

The output should be `Xa`, but instead we get `XA`. This is fixable by having OneShot track more event state. Probably the best way to accomplish this would be to make it a subscriber to a global event queue. That would be a fairly big change, affecting several plugins, including TapDance, and might solve other plugin-interaction issues. It could also be fixed by having OneShot store more state on its own. Depending on the perceived severity of this individual bug, I could do the latter before eventually introducing the more general solution.

gedankenexperimenter commented:
> Do you have an alternate name you’d propose for Key_Unbound?

For some reason, I didn't get a notification when you sent this, so I didn't see it right away; sorry. I think `Key_Inactive` or `Key_Idle` better describes the meaning in the context of the `live_keys[]` array.  "Unbound" strikes me as more appropriate for an entry in the keymap than the live keyboard state.

gedankenexperimenter commented:
I had an idea this morning for a way to fix the OneShot/TapDance issue, and got it working. It introduces one more event handler hook: `afterReportingState()`, which allows a plugin to react to an event after that event's report is sent to the host, but before any other events that are processed during the same cycle. It's not a perfect solution, because of plugin registration order dependencies (if two plugins want to react to one event after the fact, it's possible to get different results based on registration order), but right now only OneShot has a need for it.

I can add the fix to this PR once I get the commit history for it cleaned up, or make a separate one after this gets merged, whichever you guys prefer.

algernon commented:
Regarding `Key_Unbound` and the alternatives: I think I like `Key_Unbound` or `Key_Idle` best, but `Key_Inactive` isn't far off, either. Any of those work for me, so I'll leave that up to @obra.

> I can add the fix to this PR once I get the commit history for it cleaned up, or make a separate one after this gets merged, whichever you guys prefer.

Separate PR, please, I want to merge this ASAP.

With these said, once the final name `Key_Unbound` is solved, I think this should go in right after.

gedankenexperimenter commented:
> Separate PR, please, I want to merge this ASAP.

Certainly, sir! I'll have a batch of small PRs to submit once this is merged, and I'll include that as one.

> With these said, once the final name `Key_Unbound` is solved, I think this should go in right after.

`Key_Unbound` has been replaced with `Key_Inactive`, as discussed. A couple of glossary entries have been added, as well. I suspect you might want to make some adjustments, as I found myself uncertain as to how to best write and organize those entries, but at least they're in there, and can be improved upon later.

algernon commented:
This took way longer to merge than I would've wished, but it's in! Thank you a **lot** for the tremendous effort you put in, and for the patience aswell.

gedankenexperimenter commented:
You give me more credit than I deserve for patience—thank you. Now for all the stuff I've been waiting on this one for! 😉
